### ISSUE TITLE
Enabling DEAL_II_WITH_NETCDF causes the build to fail: 'netcdfcpp.h' file not found

### ISSUE THREAD

**Issue by yurivict** [dealii/dealii#7263]
```
/usr/ports/math/dealii/work/dealii-9.0.0/source/grid/grid_in.cc:36:10: fatal error: 'netcdfcpp.h' file not found
#include <netcdfcpp.h>
         ^~~~~~~~~~~~~
```

The packages ```netcdf``` and ```netcdf-cxx``` don't install ```netcdfcpp.h```.

**Comment by tamiko** [comment]
How did you enable netcdf support? Can you please attach the file `detailed.log` and your cmake command line invocation?

Our `FindNETCDF.cmake` contains:
```
 28 DEAL_II_FIND_PATH(NETCDF_INCLUDE_DIR netcdfcpp.h                                 
 29   HINTS ${NETCDF_DIR}                                                            
 30   PATH_SUFFIXES netcdf include                                                   
 31   )
```
so it is a bit puzzling why this doesn't end up in the final list of include directories.

**Comment by yurivict** [comment]
I enabled it by adding ```-DDEAL_II_WITH_NETCDF=ON```.
There is no such file ```netcdfcpp.h``` in the standard locations, only ```/usr/local/include/vtk-6.2/vtknetcdf/include/netcdfcpp.h``` installed by ```vtk-6.2```.
Where is it supposed to come from?

**Comment by davydden** [comment]
> The packages netcdf and netcdf-cxx don't install netcdfcpp.h.

I don’t how those are configured, but this header should be there. That’s how we configure deal in Spack 
https://github.com/davydden/spack/blob/ca97d8c928eaa84eb25c3ecf2e4c86856458ac32/var/spack/repos/builtin/packages/dealii/package.py

**Comment by yurivict** [comment]
What package installs ```netcdfcpp.h```?

**Comment by jppelteret** [comment]
On my machine it is `netcdf-cxx`:
```sh
$ls <path>/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/netcdf-cxx-4.2-tagcl5v6ublio3db6jganwn25hyb4fxy/include
ncvalues.h    netcdf.hh    netcdfcpp.h
```

**Comment by yurivict** [comment]
The latest version of ```netcdf-cxx``` is ```4.3.0```, and it doesn't install ```netcdfcpp.h```.
The complete list of files:
```
$ pkg info -l netcdf-cxx
netcdf-cxx-4.3.0_1:
	/usr/local/bin/ncxx4-config
	/usr/local/include/ncAtt.h
	/usr/local/include/ncByte.h
	/usr/local/include/ncChar.h
	/usr/local/include/ncCheck.h
	/usr/local/include/ncCompoundType.h
	/usr/local/include/ncDim.h
	/usr/local/include/ncDouble.h
	/usr/local/include/ncEnumType.h
	/usr/local/include/ncException.h
	/usr/local/include/ncFile.h
	/usr/local/include/ncFloat.h
	/usr/local/include/ncGroup.h
	/usr/local/include/ncGroupAtt.h
	/usr/local/include/ncInt.h
	/usr/local/include/ncInt64.h
	/usr/local/include/ncOpaqueType.h
	/usr/local/include/ncShort.h
	/usr/local/include/ncString.h
	/usr/local/include/ncType.h
	/usr/local/include/ncUbyte.h
	/usr/local/include/ncUint.h
	/usr/local/include/ncUint64.h
	/usr/local/include/ncUshort.h
	/usr/local/include/ncVar.h
	/usr/local/include/ncVarAtt.h
	/usr/local/include/ncVlenType.h
	/usr/local/include/netcdf
	/usr/local/lib/libnetcdf_c++4.a
	/usr/local/lib/libnetcdf_c++4.so
	/usr/local/lib/libnetcdf_c++4.so.1
	/usr/local/lib/libnetcdf_c++4.so.1.0.3
	/usr/local/libdata/pkgconfig/netcdf-cxx4.pc
	/usr/local/share/licenses/netcdf-cxx-4.3.0_1/LICENSE
	/usr/local/share/licenses/netcdf-cxx-4.3.0_1/UCAR_UNIDATA
	/usr/local/share/licenses/netcdf-cxx-4.3.0_1/catalog.mk
```

**Comment by jppelteret** [comment]
Oh, well that's frustrating. I don't have any suggestions since my version of netcdf is clearly quite old (in spack, up to 4.6.1 is listed) and I don't use it. Maybe someone else (who uses it?) would have some suggestions. Perhaps we need to update how its `#include`d in the library.

**Comment by jppelteret** [comment]
Oh, I made an error (I was looking at netcdf instead of the netcdf-cxx package). Spack also has a netcdf-cxx4 package that packages version 4.3.0 (which, like you state, appears to be the most recent version). This is just confusing. Clearly our spack package references the older netcdf-cxx that only supplies version 4.2.

**Comment by davydden** [comment]
indeed, in Spack we build `deal.II` against Netcdf-cxx 4.2, according to package description those are deprecated unmaintained C++ bindings, probably that's the reason why there is a separate `netcdf-cxx4` package in Spack. Note that the most recent version name the library as `libnetcdf_c++4.so` whereas the old one is `libnetcdf_c++.so`. So it seems they are really meant as different packages.

@tamiko I guess we should see if we can support the current versions 4.3.0.

**Comment by yurivict** [comment]
This happens all the time. Some package changes interface in a major way, and since there are too many other packages depending on it they make it installable in parallel, and you would have multiple packages. Individual packages like ```deal-II``` should try to stay with the latest version.

**Comment by tamiko** [comment]
Our netcdf interface is ancient and has never been ported to the new include header layout / library interface (in at least the last 10 years if I recall correctly).

@yurivict Can you please show me the corresponding `detailed.log`. Let me put it this way: The configure step should have never been successful if the header is not found.

**Comment by tamiko** [comment]
@yurivict Interesting. Do you set `NETCDF_INCLUDE_DIR`, or `NETCDF_INCLUDE_DIRS` by hand?
Further, what is your configure output when running cmake saying about netcdf? It should look like this if the header is not available:
```
-- Include /home/tamiko/workspace/dealii/cmake/configure/configure_netcdf.cmake
-- NETCDF_INCLUDE_DIR not found! Call:
--     FIND_PATH(NETCDF_INCLUDE_DIR netcdfcpp.h HINTS PATH_SUFFIXES netcdf include)
-- Found NETCDF_CPLUSPLUS_LIBRARY
-- Found NETCDF_C_LIBRARY
--   NETCDF_LIBRARIES: /usr/lib64/libnetcdf_c++.so;/usr/lib64/libnetcdf.so
--   NETCDF_INCLUDE_DIRS: *** Required variable "NETCDF_INCLUDE_DIR" set to NOTFOUND ***
-- Could NOT find NETCDF
```

**Comment by yurivict** [comment]
I set
```-DNETCDF_INCLUDE_DIR:STRING=${LOCALBASE}/include -DNETCDF_CPLUSPLUS_LIBRARY:STRING=${LOCALBASE}/lib/libnetcdf_c++4.so```

configure prints:
```
--
-- Include /usr/ports/math/deal.ii/work/dealii-9.0.0/cmake/configure/configure_netcdf.cmake
-- Found NETCDF_INCLUDE_DIR
-- Found NETCDF_CPLUSPLUS_LIBRARY
-- Found NETCDF_C_LIBRARY
--   NETCDF_LIBRARIES: /usr/local/lib/libnetcdf_c++4.so;/usr/local/lib/libnetcdf.so
--   NETCDF_INCLUDE_DIRS: /usr/local/include
-- Found NETCDF
-- DEAL_II_WITH_NETCDF successfully set up with external dependencies.
--
```

**Comment by tamiko** [comment]
@yurivict The `-DNETCDF_INCLUDE_DIR:STRING=${LOCALBASE}/include` is the problem. By setting the include directory you are bypassing the internal configure check whether the header exists.

**Comment by tamiko** [comment]
Closing in favor of https://github.com/dealii/dealii/issues/5570
